rfcs/20230621-tf-api-deprecation.md

```diff
@@ -0,0 +1,101 @@
+* As appropriate, examples (e.g. experimental Colab notebooks) are created demonstrating the replacement of any deprecated modules or methods.
+* The API is covered by a publicly available _strict mode_ which users can optionally enable to convert warnings to failures, with error messages that guide them to replacements.
+    * Strict mode is a globally applied state, switched on with `tf.exprimental.enable_strict_mode()`. 
+    * Once enabled, ignorable runtime warnings for deprecated APIs will instead be replaced with errors, detailing the deprecation and suggested alternatives.
+* Alongside release in TF Nightly, messaging is sent through appropriate communication channels to inform of the upcoming deprecation, rationale, and available substitutes.
+    * Depending on the nature and scope of deprecation, this may be done at the individual API level or in interrelated block(s).
+    * The primary channel for this will be the TensorFlow release notes. Each release shall have a dedicated, prominent Deprecation section (à la [NumPy](https://numpy.org/doc/stable/release/1.24.0-notes.html#deprecations)) where announcements are made.
+
+### Removal
+
+Removal is substantial, but straightforward. Code is deleted. Should the appropriate precursor steps be taken as specified, the groundwork is set for an as-smooth-as-possible introduction of backwards incompatibility.
+
+### Compatibility
+
+These are relevant for API removal.
+
+#### SavedModels
+
+SavedModels written with one version of TensorFlow can be loaded and evaluated with a later version of TensorFlow within the same major release. This is an end-to-end guarantee: the model can be loaded, the functions registered, …, all the way through runtime execution.
+
+#### GraphDefs and Checkpoints
+
+GraphDef currently has versioning and a given release of TensorFlow supports a range of versions of the GraphDef.  Note that the GraphDef version mechanism is separate from the TensorFlow version. 
+
+TensorFlow will provide a **1 year** backward compatibility guarantee for GraphDefs. This is to support restoring graphs created with the immediately previous version of TensorFlow. 
+
+Forward compatibility support will also be guaranteed for 3 weeks, which implies creating a GraphDef in a newer version of TensorFlow and loading it into an older version of TensorFlow [standing policy].
+
+All the same applies to checkpoints. 
+
+### Engineering Impact
+* Package size and import time should gradually decrease as APIs removed, although likely not substantially.
+
+### User Impact
+Users will begin to see runtime warnings for usage of deprecated APIs, and, upon removal ~12 months from deprecation, will be expected to either:
+* Replace its usage with the recommended alternative, OR 
+* Pin to the last version still including the API to preserve its functionality.
+
+## Questions and Discussion Topics
+
+* Should we consider a more aggressive removal timeline than 1 year post-deprecation (e.g. 6 months)?
```

Comment:
I would say no, due to security: TF no longer patches all releases in a year, so forcing people to pin to a release from 6 months ago will result in weakening of their security posture. (we never promised more than 1 year of patching, so having users pin to a version older than a year is a risk that we never covered)